jax/_src/core.py

```diff
@@ -146,13 +147,26 @@ def __str__(self):
   def pretty_print(self, *, source_info=False, print_shapes=True,
                    custom_pp_eqn_rules=True, name_stack=False,
                    print_effects: bool = False, **kw):
-    doc = pp_jaxpr(self, JaxprPpContext(),
-                   JaxprPpSettings(source_info=source_info,
-                                   print_shapes=print_shapes,
-                                   custom_pp_eqn_rules=custom_pp_eqn_rules,
-                                   name_stack=name_stack,
-                                   print_effects=print_effects))
-    return doc.format(**kw)
+    context = JaxprPpContext()
+    settings = JaxprPpSettings(source_info=source_info,
+                    print_shapes=print_shapes,
+                    custom_pp_eqn_rules=custom_pp_eqn_rules,
+                    name_stack=name_stack,
+                    print_effects=print_effects)
+
+    jaxpr_usages = Counter(subjaxprs(self, recursive=True))
+    jaxpr_usages[self] += 1
+    docs = []
+    top_level_jaxprs: dict[Jaxpr, str]
+    context.top_level_jaxprs = top_level_jaxprs = {}
+    for jaxpr, c in jaxpr_usages.items():
+      if c == 1:
+        continue
+      name = "jaxpr_" + _encode_digits_alphabetic(len(top_level_jaxprs))
```

Comment:
Wouldn't it be better if we used, say, a uniquified version of the name at one of the call sites?

Reply:
Done, PTAL.